### DISCUSSION TITLE
DependsOn BelongsTo not updating UI despite updated QueryBuilder

### DISCUSSION THREAD

**Discussion by ghzb** [laravel/nova-issues#4944] [Ideas & Feature Requests] [answered]
- Laravel Version: 8.65
- Nova Version: 4.16
- PHP Version: 8.1
- Database Driver & Version: MySql 8
- Operating System and Version: macOS 12.6
- Browser type and version: Chrome 106.0.5249.119
- Reproduction Repository: n/a

### Description:
Have a simple belongsTo relationship on a model called `project` which depends on the selection of another belongsTo field. 
Project belongsTo Company
Project belongsTo User

When creating a new `project` the code below works flawlessly.  However, when I'm editing/updating a `project` the belongsTo for the User field does not update the list of available users. 

When using xDebug,  I set a breakpoint in the `relatableUsers` method on the line `return $query->where....` then stepped one step further to inspect the value of the query builder.  The querybuilder was 100% updated because something like `$query->toSql()` would return an updated version of the query. And executing the query would return the expected results.

IMO everything is working as expected, but the UI is not getting updated.

```php
class Project extends Resource
{
    public static function relatableUsers(NovaRequest $request, $query, BelongsTo $field)
    {
        $meta = collect($field->meta());
        if ($meta->has('company_id')) {
            return $query->where('company_id', $meta->get('company_id'));
        }
    }

    //....
    public function fields(Request $request)
    {
        return [
            BelongsTo::make('Contact', 'creator', NovaUser::class)
            ->dependsOn('company', function (BelongsTo $field, NovaRequest $request, FormData $formData) {
                $companyId = $formData->resource(NovaCompany::uriKey(), $formData->company);
                $field->visible = isset($companyId);
                $field->withMeta(['company_id' => $companyId]);
            })
        ];
    }
}
```

Screenshot of when xDebug values triggered on each request.  Shows the new value is updated when the other field changes.   Yet the UI does not update. 
![image](https://user-images.githubusercontent.com/65547234/196546473-2dfe4093-1b74-4b51-8c30-74531eceb52b.png)

Screenshot showing that the `BelongsTo` field makes a static call to the `relatableUsers` method passing the request, the query, and the field.   The field is how the meta data is accessed on the resource.  Yet the UI does not change.

![image](https://user-images.githubusercontent.com/65547234/196547005-143996a1-8709-47d4-aa2d-b5e46f550125.png)

Also should mention that some relatable fields make two separate requests to the API endpoints, while others only make one.  When there are two requests,  we use the session to hold onto the meta value, but the UI won't update when there is only one request.
```
if ($request->input('company')) {
    session()->put($key, $request->input('company'));
}

if (session()->has($key)) {
    return $query->where('company_id', session($key));
}
```

### Detailed steps to reproduce the issue on a fresh Nova installation:
* Create three models (project, user, company)
* Project belongsTo Company
* Project belongsTo User
* Company hasMany Project
* User hasMany Project
* Company hasMany User
* User belongsTo Company

**Comment by crynobone** [comment, marked as answer]
Implemented in 4.19.0

You can directly handle relatableQuery from `BelongsTo` instance. <https://nova.laravel.com/docs/4.0/resources/relationships.html#relatable-query-filtering>
